mava/networks/retention.py

```diff
@@ -0,0 +1,324 @@
+        # Head size
+        self.head_size = self.embed_dim // self.n_head
+
+        # Decay kappa for each head
+        self.decay_kappas = 1 - jnp.exp(
+            jnp.linspace(jnp.log(1 / 32), jnp.log(1 / 512), self.n_head)
+        )
+        self.decay_kappas = self.decay_kappas * self.decay_scaling_factor
+
+        # Initialize the weights and group norm
+        self.w_g = self.param(
+            "w_g",
+            nn.initializers.normal(stddev=1 / self.embed_dim),
+            (self.embed_dim, self.head_size),
+        )
+        self.w_o = self.param(
+            "w_o",
+            nn.initializers.normal(stddev=1 / self.embed_dim),
+            (self.head_size, self.embed_dim),
+        )
+        self.group_norm = nn.GroupNorm(num_groups=self.n_head)
+
+        # Initialize the retention mechanisms
+        self.retentions = [
+            SimpleRetention(
+                self.embed_dim,
+                self.head_size,
+                self.n_agents,
+                self.full_self_retention,
+                decay_kappa,
+                self.memory_config,
+            )
+            for decay_kappa in self.decay_kappas
+        ]
+
+        # Create an instance of the positional encoding
+        self.pe = PositionalEncoding(self.memory_config, self.embed_dim)
+
+    def __call__(
+        self,
+        key: Array,
+        query: Array,
+        value: Array,
+        hstate: Array,
+        dones: Array,
+        step_count: Array,
+    ) -> Tuple[Array, Array]:
+        """Chunkwise (default) representation of the multi-scale retention mechanism"""
+        B, C, _ = value.shape
+
+        # Positional encoding of the current step
+        key, query, value = self.pe(key, query, value, step_count)
+
+        # Per head retention
+        ret_output = jnp.zeros((B, C, self.head_size), dtype=value.dtype)
+        hs = jnp.copy(hstate)
+        for head in range(self.n_head):
+            y, new_hs = self.retentions[head](key, query, value, hstate[:, head], dones)
+            ret_output = ret_output.at[
+                :, :, self.head_size * head : self.head_size * (head + 1)
+            ].set(y)
+            hs = hs.at[:, head, :, :].set(new_hs)
+
+        ret_output = self.group_norm(ret_output.reshape(-1, self.head_size)).reshape(
+            ret_output.shape
+        )
+
+        x = key
+        output = (jax.nn.swish(x @ self.w_g) * ret_output) @ self.w_o
+        return output, hs
+
+    def recurrent(
+        self, key_n: Array, query_n: Array, value_n: Array, hstate: Array, step_count: Array
+    ) -> Tuple[Array, Array]:
+        """Recurrent representation of the multi-scale retention mechanism"""
+        B, S, _ = value_n.shape
+
+        # Positional encoding of the current step
+        key_n, query_n, value_n = self.pe(key_n, query_n, value_n, step_count)
+
+        ret_output = jnp.zeros((B, S, self.head_size), dtype=value_n.dtype)
+        hs = jnp.zeros_like(hstate)  # hidden state per head
```

Comment:
Similarly here.

Reply:
maybe just for consistency i will use zero_like for both cases, what do you think?